tests/chainermn_tests/links_tests/test_n_step_rnn.py

```diff
@@ -39,93 +39,109 @@ def __call__(self, xs, ts):
-        else:
-            self.communicator = chainermn.create_communicator('naive')
+    if communicator.size < 2:
+        pytest.skip('This test is for multinode only')
 
-        if self.communicator.size < 2:
-            pytest.skip('This test is for multinode only')
+    rank_next = communicator.rank + 1
+    rank_prev = communicator.rank - 1
 
-        self.rank_next = self.communicator.rank + 1
-        self.rank_prev = self.communicator.rank - 1
+    if rank_prev < 0:
+        rank_prev = None
 
-        if self.rank_prev < 0:
-            self.rank_prev = None
+    if rank_next >= communicator.size:
+        rank_next = None
 
-        if self.rank_next >= self.communicator.size:
-            self.rank_next = None
+    return communicator, rank_prev, rank_next
 
-    def check_homogeneous_rnn(self, gpu):
-        self.setup(gpu)
 
-        n, n_vocab, l = 100, 8, 10
-        # Number of model parameters are same among processes.
-        n_hid = 2
+def check_homogeneous_rnn(gpu, dtype):
+    communicator, rank_prev, rank_next = setup_communicator(gpu=gpu)
 
+    n, n_vocab, l = 100, 8, 10
+    # Number of model parameters are same among processes.
+    n_hid = 2
+    with chainer.using_config('dtype', dtype):
         X = [np.random.randint(
-            0, n_vocab, size=np.random.randint(l // 2, l + 1), dtype=np.int32)
+            0, n_vocab, size=np.random.randint(l // 2, l + 1),
+            dtype=np.int32)
             for _ in range(n)]
-        Y = (np.random.rand(n) * 2).astype(np.float32)
+        Y = (np.random.rand(n) * 2).astype(dtype)
         model = Model(
-            n_vocab, n_hid, self.communicator, self.rank_next, self.rank_prev)
+            n_vocab, n_hid, communicator, rank_next,
+            rank_prev)
 
         if gpu:
             model.to_device(cupy.cuda.Device())
-            X = [chainer.cuda.to_gpu(x) for x in X]
-            Y = chainer.cuda.to_gpu(Y)
+            X = [chainer.backends.cuda.to_gpu(x) for x in X]
+            Y = chainer.backends.cuda.to_gpu(Y)
 
         for i in range(n):
             err = model(X[i:i + 1], Y[i:i + 1])
             err.backward()
 
         # Check if backprop finishes without deadlock.
-        self.assertTrue(True)
+        assert True
+
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
-    def test_homogeneous_rnn_cpu(self):
-        self.check_homogeneous_rnn(False)
+@pytest.mark.filterwarnings('ignore::DeprecationWarning')
+@pytest.mark.parametrize('dtype', [np.float16, np.float32])
+def test_homogeneous_rnn_cpu(dtype):
+    check_homogeneous_rnn(False, dtype)
 
-    @chainer.testing.attr.gpu
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
-    def test_homogeneous_rnn_gpu(self):
-        self.check_homogeneous_rnn(True)
 
-    def check_heterogeneous_rnn(self, gpu):
-        self.setup(gpu)
+@chainer.testing.attr.gpu
+@pytest.mark.filterwarnings('ignore::DeprecationWarning')
```

Comment:
Why this filter is needed?

Reply:
These are legacy filters from https://github.com/chainer/chainer/commit/bbbcf9358ea1b5cc156a360853eb0bbd7beb89bf
to surpass the deprecated `add_link` https://github.com/chainer/chainer/blob/76a4aeef7ef86be8c273995b34338f89d00273f7/chainer/link.py#L920
The deprecated mark is removed, so I also remove the filter. There is another file affected 
https://github.com/chainer/chainer/blob/307a858b053e86b140fa71432fa60950dc83bee6/tests/chainermn_tests/links_tests/test_multi_node_chain_list.py#L314
I will open an issue for it